### PR TITLE
fix: invalid validation of large offset case

### PR DESCRIPTION
**Pull request by prm-dan**
PRO-5680

This was existing code that I should have updated.

I also did some extra refactoring.

TESTING=ran spec

### CODE REVIEW

File: lib/promoted/ruby/client/pager.rb [hunk 1]
```diff
@@ -13,34 +13,43 @@ def initialize(message, default_insertions_page)
...
                 }
               end
 
-              offset = [0, paging[:offset]].max
-              retrieval_insertion_offset = [0, retrieval_insertion_offset].max
-              index = [0, offset - retrieval_insertion_offset].max
+              offset, retrieval_insertion_offset, index = _sanitize_offsets(retrieval_insertion_offset, paging)
+
+              begin
+                _validate_paging(insertions, retrieval_insertion_offset, offset, index)
```
Comment (prm-dan) [inline lib/promoted/ruby/client/pager.rb:48]: The validation logic is easier if we are looking at the sanitized offsets.

File: lib/promoted/ruby/client/pager.rb [hunk 2]
```diff
@@ -13,34 +13,43 @@ def initialize(message, default_insertions_page)
...
+              retrieval_insertion_offset = [0, retrieval_insertion_offset].max
+              index = [0, offset - retrieval_insertion_offset].max
+              return [offset, retrieval_insertion_offset, index]
+            end
+
+            def _validate_paging(insertions, retrieval_insertion_offset, offset, index)
+              if offset < retrieval_insertion_offset
+                raise InvalidPagingError.new("Invalid page offset (retrieval_insertion_offset #{retrieval_insertion_offset}, offset #{offset})", [])
+              end
+              if index >= insertions.length
```
Comment (prm-dan) [inline lib/promoted/ruby/client/pager.rb:32]: We need to use index instead here (since index is relative to the insertions list).

**Review by prm-james-hill — APPROVED** [review]

File: lib/promoted/ruby/client/pager.rb (hunk 2 above)
Comment (prm-james-hill) [reply lib/promoted/ruby/client/pager.rb:32]: Yeah, when I was writing this logic for the .NET SDK I drafted a lot of comments about what each index precisely was,  but it was really hard to explain

Comment (prm-dan) [reply lib/promoted/ruby/client/pager.rb:32]: Longer-term, I'd guess we'll rip out some of this validation logic.  It moves a corner case that we could handle in the SDK to a problem for the caller.

**Pull request merged**